gluten-ras/common/src/test/scala/org/apache/gluten/ras/PropertySuite.scala

```diff
@@ -37,6 +38,30 @@ abstract class PropertySuite extends AnyFunSuite {
 
   protected def conf: RasConfig
 
+  test("Group memo - cache") {
+    val ras =
+      Ras[TestNode](
+        PlanModelImpl,
+        CostModelImpl,
+        MetadataModelImpl,
+        NodeTypePropertyModelWithOutEnforcerRules,
+        ExplainImpl,
+        RasRule.Factory.none())
+        .withNewConfig(_ => conf)
+
+    val memo = Memo(ras)
+
+    memo.memorize(ras, PassNodeType(1, PassNodeType(1, PassNodeType(1, TypedLeaf(TypeA, 1)))))
+    val leafGroup = memo.memorize(ras, TypedLeaf(TypeA, 1))
+    memo
+      .openFor(leafGroup.clusterKey())
+      .memorize(ras, TypedLeaf(TypeB, 1))
+    memo.memorize(ras, PassNodeType(1, PassNodeType(1, PassNodeType(1, TypedLeaf(TypeB, 1)))))
+    val state = memo.newState()
+    assert(state.allClusters().size == 4)
+    assert(state.getGroupCount() == 11)
+  }
```

Comment:
Memo dump of this case:

Before:
![graphviz](https://github.com/apache/incubator-gluten/assets/11284395/3e426942-e939-46c7-9595-c9401ed9d26e)

After:
![graphviz (1)](https://github.com/apache/incubator-gluten/assets/11284395/7c2dec76-0cd4-4845-8ab3-5481aa11339b)